File: basic/data_structure/python/BinaryTree.py
class TreeNode(object):
    def __init__(self, val):
        self.val = val
        self.left = self.right = None

class Traversal(object):
    def __init__(self):
        self.result = list()

    def preorder(self, root):
        if root:
            self.result.append(root.val)
            self.preorder(root.left)
            self.preorder(root.right)

    def inorder(self, root):
        if root:
            self.inorder(root.left)
            self.result.append(root.val)     
            self.inorder(root.right)

    def postorder(self, root):
        if root:
            self.postorder(root.left)
            self.postorder(root.right)
            self.result.append(root.val)

File: basic/data_structure/java/BinaryTree.java
import java.util.ArrayList;
import java.util.List;

class TreeNode<T extends Comparable<T>> {
    public T val;
    public TreeNode<T> left, right;

    public TreeNode(T val) {
        this.val = val;
        this.left = this.right = null;
    }
}

class Traversal<T extends Comparable<T>> {

    public List<T> preorder(TreeNode<T> root) {
        List<T> result = new ArrayList<>();
        recursivePreorder(root, result);
        return result;
    }

    private void recursivePreorder(TreeNode<T> node, List<T> result) {
        if (node != null) {
            result.add(node.val);
            recursivePreorder(node.left, result);
            recursivePreorder(node.right, result);
        }
    }

    public List<T> inorder(TreeNode<T> root) {
        List<T> result = new ArrayList<>();
        recursiveInorder(root, result);
        return result;
    }

    private void recursiveInorder(TreeNode<T> node, List<T> result) {
        if (node != null) {
            recursivePreorder(node.left, result);
            result.add(node.val);
            recursivePreorder(node.right, result);
        }
    }

    public List<T> postorder(TreeNode<T> root) {
        List<T> result = new ArrayList<>();
        recursivePostorder(root, result);
        return result;
    }

    private void recursivePostorder(TreeNode<T> node, List<T> result) {
        if (node != null) {
            recursivePreorder(node.left, result);    
            recursivePreorder(node.right, result);
            result.add(node.val);
        }
    }
}